src/Cli/Microsoft.DotNet.FileBasedPrograms/FileLevelDirectiveHelpers.cs

```diff
@@ -0,0 +1,619 @@
+
+        public static new Package? Parse(in ParseContext context)
+        {
+            if (ParseOptionalTwoParts(context, separator: '@') is not var (packageName, packageVersion))
+            {
+                return null;
+            }
+
+            return new Package(context.Info)
+            {
+                Name = packageName,
+                Version = packageVersion,
+            };
+        }
+
+        public override string ToString() => Version is null ? $"#:package {Name}" : $"#:package {Name}@{Version}";
+    }
+
+    /// <summary>
+    /// <c>#:project</c> directive.
+    /// </summary>
+    public sealed class Project(in ParseInfo info) : Named(info)
+    {
+        public static new Project? Parse(in ParseContext context)
+        {
+            var directiveText = context.DirectiveText;
+            if (directiveText.IsWhiteSpace())
+            {
+                string directiveKind = context.DirectiveKind;
+                return context.Diagnostics.AddError<Project?>(context.SourceFile, context.Info.Span, string.Format(FileBasedProgramsResources.MissingDirectiveName, directiveKind));
+            }
+
+            try
+            {
+                // If the path is a directory like '../lib', transform it to a project file path like '../lib/lib.csproj'.
+                // Also normalize blackslashes to forward slashes to ensure the directive works on all platforms.
+                var sourceDirectory = Path.GetDirectoryName(context.SourceFile.Path) ?? ".";
+                var resolvedProjectPath = Path.Combine(sourceDirectory, directiveText.Replace('\\', '/'));
+                if (Directory.Exists(resolvedProjectPath))
+                {
+                    var fullFilePath = GetProjectFileFromDirectory(resolvedProjectPath).FullName;
+
+                    // Keep a relative path only if the original directive was a relative path.
+                    directiveText = ExternalHelpers.IsPathFullyQualified(directiveText)
+                        ? fullFilePath
+                        : ExternalHelpers.GetRelativePath(relativeTo: sourceDirectory, fullFilePath);
+                }
+                else if (!File.Exists(resolvedProjectPath))
+                {
+                    throw new GracefulException(FileBasedProgramsResources.CouldNotFindProjectOrDirectory, resolvedProjectPath);
+                }
+            }
+            catch (GracefulException e)
+            {
+                context.Diagnostics.AddError(context.SourceFile, context.Info.Span, string.Format(FileBasedProgramsResources.InvalidProjectDirective, e.Message), e);
+            }
+
+            return new Project(context.Info)
+            {
+                Name = directiveText,
+            };
+        }
+
+        public Project WithName(string name)
+        {
+            return new Project(Info) { Name = name };
+        }
+
+        public static FileInfo GetProjectFileFromDirectory(string projectDirectory)
```

Comment:
This helper was copied from `MsbuildProject.GetProjectFileFromDirectory`. I think if we fix the issue with `MSBuildUtilities.cs`, then, we should also be able to adjust things so that we can reuse the CLI's copy of the method.